src/main.rs

```diff
@@ -34,52 +36,93 @@ header! { (XCachetToken, "X-Cachet-Token") => [String] }
-                .header(XCachetToken(match get_bearer_token(req) {
-                    Ok(token) => token,
+            let http_client = reqwest::Client::new();
+
+            for (component, status) in components {
+
+                //create map for json of cachet api call
+                let mut map = HashMap::new();
+                map.insert("status", status);
+
+                match get_bearer_token(req.clone()) {
+                    Ok(token) => {//send http put to cachet
+                        match http_client.put(&format!(
+                            "{}/api/v1/components/{}",
+                            match std::env::var("CACHET_BASE_URL") {
+                                Ok(val) => val,
+                                Err(_) => String::from(include_str!("cachet_endpoint.txt")),
+                            },
+                            component
+                        ))
+                            // read "Authorization: Bearer" token into "X-Cachet_Token"
+                            .header(XCachetToken(token)).json(&map).send() {
+                            Ok(res) => {
+                                responses.push(CachetResponse {
+                                    http_status: res.status().as_u16(),
+                                    status: CachetAnnotation {
+                                        component,
+                                        severity: status,
+                                    }
+                                })
+                            },
+                            Err(err) => {
+                                error!("Could not contact the cachet API: {}", err);
+                                responses.push(CachetResponse {
+                                    http_status: http::StatusCode::INTERNAL_SERVER_ERROR.as_u16(),
+                                    status: CachetAnnotation {
+                                        component,
+                                        severity: status,
+                                    }
+                                })
+                            }
+                        }
+                    },
                     Err(err) => {
+                        responses.push(CachetResponse {
+                            http_status: http::StatusCode::UNAUTHORIZED.as_u16(),
+                            status: CachetAnnotation {
+                                component,
+                                severity: status,
+                            }
+                        });
                         error!("Unable to find bearer token in the \"Authorization\" header: {}", err);
-                        return Ok(HttpResponse::new(http::StatusCode::UNAUTHORIZED));
                     }
                 }
-                )).json(&map).send() {
-                Ok(res) => res,
+            };
+
+
+            let (status, body) = match serde_json::to_string(&responses) {
```

Comment:
Will this iterate over the vector of responses? And if it's OK it'll do nothing, otherwise it'll throw the error, then if it hasn't thrown an error by line 124, it'll return OK ?

Reply:
So to_string() will return a "Result", which is an enum that either encapsulates a value in Ok(val), or an error in it's Err(error) variant. The match checks which of those two cases it goes to and then puts either (200, body) or (500, error text) in the (status, body) tuple.

Unless something goes really wrong, we always should be able to serialize those objects, so the error case ***should*** never occur, but it's nicer to have it covered.

For an example what I get in insomnia when hitting that endpoint, here the request:
```
{
  "version": "4",
  "groupKey": "<string>",   
  "status": "resolved",
  "receiver": "<string>",
  "groupLabels": {},
  "commonLabels": {},
  "commonAnnotations": {},
  "externalURL": "<string>", 
  "alerts": [
    {
      "status": "resolved",
      "labels": {},
      "annotations": {
				"component": "3",
				"severity": "2"
			},
      "startsAt": "<rfc3339>",
      "endsAt": "<rfc3339>",
      "generatorURL": "<string>"
    },
    {
      "status": "resolved",
      "labels": {},
      "annotations": {
				"component": "3",
				"severity": "3"
			},
      "startsAt": "<rfc3339>",
      "endsAt": "<rfc3339>",
      "generatorURL": "<string>"
    },
    {
      "status": "firing",
      "labels": {},
      "annotations": {
				"component": "4",
				"severity": "2"
			},
      "startsAt": "<rfc3339>",
      "endsAt": "<rfc3339>",
      "generatorURL": "<string>"
    },
    {
      "status": "firing",
      "labels": {},
      "annotations": {
				"component": "6",
				"severity": "1"
			},
      "startsAt": "<rfc3339>",
      "endsAt": "<rfc3339>",
      "generatorURL": "<string>"
    }
  ]
}
```
and here the response:
```
[
	{
		"httpStatus": 200,
		"status": {
			"component": "3",
			"severity": "1"
		}
	},
	{
		"httpStatus": 404,
		"status": {
			"component": "4",
			"severity": "2"
		}
	},
	{
		"httpStatus": 200,
		"status": {
			"component": "6",
			"severity": "1"
		}
	}
]
```
This also shows quite nicely what happens when a component on the cachet side does not exist, we get a 404 for that component.